### ISSUE TITLE
All Service Checks are Late

### ISSUE THREAD

**Issue by thend20** [Icinga/icinga2#5011]
I have a single master setup with 2000 servers, and 17000 services in total being monitored.  Up until yesterday checks where occuring regularly.  However, today I noticed there has not been a single Check Result since the approximately the same time yesterday for every service.  

Troubleshooting steps, which did not correct the problem
 - restarted all services, and even rebooted the monitoring server itself
 - removed all of the Host objects, restarted icinga2, and then added them back hoping that would revive the checks
 - "Force Check" does work; but the service is not checked again after the Force; it falls back into Late status.
 - Upgraded from 2.6.1 to 2.6.2

Questions

 - Is there a way to force *all* services to check, even just a single time? How about reschedule *all* services?
 - Is my database corrupt?  The Next Check for all of these services has a date of yesterday as well.  Its like *something* happened around 17:45 yesterday to break automated service checks.
 - I'm somewhat dependent on this monitoring, is there a specific table or tables in the DB I can empty to get this working asap?  I know this different solve the actual problem, but its solves my "now" problem
 - Let me know what you might need from me for troubleshooting

Icinga2 2.6.2
Icinga2-ido-mysql 2.6.2
Icingaweb2 2.4.0

**Comment by dnsmichi** [comment]
Are you using the icinga2 client, or another other agent? Is there a pattern for the checks being late?

You can jump onto the console and fetch late checks too. https://www.icinga.com/2016/08/11/analyse-icinga-2-problems-using-the-console-api/

In terms of the database vs. runtime state - pick a service from your DB/Web and compare its runtime state to the REST API GET request onto /v1/objects/services

Any specific system problems, such as load, disk IO, etc? Any trends you see in your graphs? (you've hopefully some self checks installed).

And to conplete with, please add your icinga2.conf, zones.conf and the output of `icinga2 --version`.

**Comment by thend20** [comment]
Currently we are running remote checks only, so there's no client/server remote execution going on (like icinga2 client nor nrpe).  Essentially these are tcp/udp checks (ssh/ping/ntp/http/ftp/named).

The time of the last check seems to match up between the API and the IDO.

As for specific problems, every single service check is late, and stopped responding around the same time yesterday.  Restarting/rebooting did not force the checks to happen on startup.  So, every single service check is stuck in a Late state with a Scheduled time in the past.

I've attached the requested configuration.  The icinga2.conf and zones.conf has never been touched; they are the default (renamed to txt for github).  Do you need any/everything in conf.d?

[version.txt](https://github.com/Icinga/icinga2/files/778282/version.txt)
[zones.txt](https://github.com/Icinga/icinga2/files/778284/zones.txt)
[icinga2.txt](https://github.com/Icinga/icinga2/files/778283/icinga2.txt)

**Comment by dnsmichi** [comment]
Please add the output of 

```
icinga2 feature list
icinga2 daemon -C
```

I don't need anything specific from the conf.d directory, unless we find a pattern here.

How are things going inside the icinga2.log file - any warning/critical messages around?

**Comment by dnsmichi** [comment]
Oh, and as suggested from the blog post, connect to the icinga2 console and run queries to fetch all latest objects.

```
# ICINGA2_API_USERNAME=root ICINGA2_API_PASSWORD=xxx icinga2 console --connect 'https://localhost:5665/'

<7> => var res = {}; for (service in get_objects(Service).filter(s => get_time() - s.last_check > 5 * s.check_interval)) { res[service.zone] += 1 }; res

<10> => get_objects(Service).filter(s => get_time() - s.last_check > 20 * s.check_interval).map(function(s) { s.__name })
```

**Comment by thend20** [comment]
```
# icinga2 feature list
Disabled features: compatlog gelf graphite influxdb opentsdb perfdata statusdata syslog
Enabled features: api checker command debuglog ido-mysql livestatus mainlog notification
# icinga2 daemon -C
information/cli: Icinga application loader (version: r2.6.2-1)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
information/ApiListener: My API identity: orwell2.pair.net
warning/ApplyRule: Apply rule 'satellite-host' (in /etc/icinga2/conf.d/satellite.conf: 29:1-29:41) for type 'Dependency' does not match anywhere!
warning/ApplyRule: Apply rule 'backup-downtime' (in /etc/icinga2/conf.d/downtimes.conf: 5:1-5:52) for type 'ScheduledDowntime' does not match anywhere!
warning/ApplyRule: Apply rule 'procs' (in /etc/icinga2/conf.d/services.conf: 85:1-85:21) for type 'Service' does not match anywhere!
warning/ApplyRule: Apply rule 'swap' (in /etc/icinga2/conf.d/services.conf: 95:1-95:20) for type 'Service' does not match anywhere!
information/ConfigItem: Instantiated 1 ApiUser.
information/ConfigItem: Instantiated 1 ApiListener.
information/ConfigItem: Instantiated 1 Zone.
information/ConfigItem: Instantiated 2 FileLoggers.
information/ConfigItem: Instantiated 1 Endpoint.
information/ConfigItem: Instantiated 2 NotificationCommands.
information/ConfigItem: Instantiated 8 Notifications.
information/ConfigItem: Instantiated 204 CheckCommands.
information/ConfigItem: Instantiated 2523 Hosts.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 2 HostGroups.
information/ConfigItem: Instantiated 5 Comments.
information/ConfigItem: Instantiated 1 UserGroup.
information/ConfigItem: Instantiated 1 User.
information/ConfigItem: Instantiated 17503 Services.
information/ConfigItem: Instantiated 3 TimePeriods.
information/ConfigItem: Instantiated 3 ServiceGroups.
information/ConfigItem: Instantiated 1 ExternalCommandListener.
information/ConfigItem: Instantiated 1 IdoMysqlConnection.
information/ConfigItem: Instantiated 1 LivestatusListener.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ConfigItem: Instantiated 1 NotificationComponent.
information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).
```

As for the logs, there is nothing in the error log; I enabled the debug log, and that mainly just shows MySQL queries.  The main log itself is mostly a stream of Ido status infrmation every 15 seconds or so.  I don't see any warning/crits/errors, at least none since I got everything setup.

Queries

```
<1> => var res = {}; for (service in get_objects(Service).filter(s => get_time() - s.last_check > 5 * s.check_interval)) { res[service.zone] += 1 }; res
{
        "" = 17503.000000
}
```

I'm not comfortable sending the results of the second query.  Just know its a list of all the hostnames and their subsequent service checks, like this:
```
 "server1.dom.tld.!ping", "server1.dom.tld!ssh",...
```

**Comment by dnsmichi** [comment]
No worries on obfuscating the information. I took that command straight from a customer setup I had been looking into on Tuesday evening, wasn't thinking about adding the "remove sensitive bits" note here. The output looks like what I was expecting, late check results.

Is there any command line (Notice level) logging found on the debug.log?

I was thinking about time issues on your box in the morning. Can you check whether ntpd is running, the local system time is correct? Are there any load issues on your box, graphs that would highlight some trends of your system behaving wonky?

**Comment by thend20** [comment]
Here's a snippet from the debug log, its mostly just these lines repeating:
```
[2017-02-18 09:03:36 -0500] notice/WorkQueue: Spawning WorkQueue threads for 'ApiListener, RelayQueue'
[2017-02-18 09:03:36 -0500] notice/ApiListener: Relaying 'event::UpdateRepository' message
[2017-02-18 09:03:36 -0500] notice/WorkQueue: #6 (ApiListener, SyncQueue) tasks: 0
[2017-02-18 09:03:36 -0500] notice/WorkQueue: #5 (ApiListener, RelayQueue) tasks: 0
[2017-02-18 09:03:37 -0500] notice/ThreadPool: Pool #1: Pending tasks: 0; Average latency: 0ms; Threads: 4; Pool utilization: 0.126021%
[2017-02-18 09:03:37 -0500] notice/WorkQueue: #7 (IdoMysqlConnection, ido-mysql) tasks: 0
[2017-02-18 09:03:37 -0500] notice/DbConnection: Updating programstatus table.
[2017-02-18 09:03:39 -0500] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 20027; Checks/s: 0
[2017-02-18 09:03:44 -0500] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 20027; Checks/s: 0
```
Yes, ntpd is running, and up to date:
```
18 Feb 09:14:15 ntpdate[32490]: adjust time server xyz offset -0.138402 sec
```

System resources are fine.   Nothing weird going on.

**Comment by dnsmichi** [comment]
That's really a tough one ... hm. One thing you could do is to remove the icinga2.state file in /var/lib/icinga2 - stop Icinga 2 beforehand, move it away, start Icinga 2. Keep in mind that historical data (acknowledgements, etc.) will be gone then.

Maybe another step for an analysis beforehand - query the REST API for just one service at the endpoint /v1/objects/services?filter="service.name==%22SERVICENAME%22", remove sensitive details and post them here. In combination with a call to `date +%s` beforehand.

**Comment by dnsmichi** [comment]
Hm, discussed that with a co-worker. One idea for further analysis - can you add a gdb backtrace from both running processes? The first one is the master process, the second one just executes commands.

```
pidof icinga2

gdb -p 1234 -batch -ex "thread apply all bt full" -ex "detach" -ex "q" > gdb_bt_main.log
gdb -p 1235 -batch -ex "thread apply all bt full" -ex "detach" -ex "q" > gdb_bt_cmd.log
```

https://docs.icinga.com/icinga2/latest/doc/module/icinga2/toc#!/icinga2/latest/doc/module/icinga2/chapter/development#development-debug-gdb-backtrace-running

**Comment by dnsmichi** [comment]
Oh, and in terms of sensitive information - those commands will print the process and its threads. There shouldn't be any variables exposed from memory, but it wouldn't hurt to look for your host names and service names not being part of it. Seems you care a lot about it - which is good :)

**Comment by dnsmichi** [comment]
Just another thought - are these processing actually running?

```
ps aux | grep icinga2
```

must return 2 processes, each time called. might be worthwhile to run it with `watch`.

```
watch -n 1 'ps aux | grep icinga2'
```

**Comment by thend20** [comment]
Hmm, I think I may have found the problem, and its a user error on my part.  Sorry to make you guys dig so much.

We have a database of server names and services, and a script which builds attempts to build the icinga2 configuration every 15 minutes, but only if something changes.  Turns out that script was reloading icinga2 every 15 minutes regardless of configuration change.  My theory is, this wasn't enough time for icinga to figure out its Last Check time was in the past, and to schedule new checks.  And, by the time it realized this, icinga2 was reloaded again...rinse/repeat.

I think this is the problem, but I only noticed it after blowing away the state file, and then checking the PID, and noticing it changed while attempting backtraces.  I disabled the cron, and checked the webUI, and noticed services started happening.  Sooo, either the state file (maybe), or fixing constant reloading of the configuration is the problem here (probably).

I'll reopen this Issue if I find the problem happening again (and I'm sure its not my config builder).

Thanks again, and sorry for the trouble.  At the very least I learned a lot about the CLI and API interfaces.

**Comment by dnsmichi** [comment]
Thanks for the kind feedback. I think we should monitor the behaviour at least, since ongoing reloads through either a script or the Director package updates shouldn't harm operations. @dgoetz came with a similar problem today, so I'll reopen this one here for further feedback and/or challenges in reproducing the issue.

**Comment by gunnarbeutner** [comment]
This might (or might not be) related to #4989.

**Comment by marcofl** [comment]
We had similar issues with 2.6.2 update which could be related. We have master-satellite-clients setup. some (observed 3) so far of 500+ icinga2 clients did not execute checks anymore. in icingaweb2 we saw "last check 22feb" and the red dot. After restaring the icinga2 client on the affected server, evreything was okay again. logs look like this on all clients:
{code}
[2017-02-27 12:01:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:06:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:11:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:16:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:21:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:26:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:31:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:36:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:41:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:46:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:51:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 12:56:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:01:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:06:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:11:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:16:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:21:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:26:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:31:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:36:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:41:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:46:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:51:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 13:56:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:01:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:06:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:11:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:16:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:21:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:26:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:31:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:36:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:41:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:46:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:51:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 14:56:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:01:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:06:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:11:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:16:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:21:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:26:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2017-02-27 15:31:34 +0100] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
{code}

@dnsmichi you want me to create a new issue with more information?

**Comment by dnsmichi** [comment]
Checking the nice level of the icinga2 process should be sufficient. Icinga Web 2 provides a dashboard view for overdue check results (that should always return empty). Or you'll deal with some custom analysis using the console and Icinga 2 DSL. There's a blog post on icinga.com/blog on using the console cli command.

I've discussed those issues with @dgoetz and as such they're related to #4989. Closing here.
